Expose Quirks Mode information in the layout data and code

julienw opened the pull request:
<!-- Please describe your changes on the following line: -->
This patch exposes the Quirks (NoQuirks/LimitedQuirks/Quirks) state to the layout subsystem.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).
Prelimary work for issue #11704.

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____
(Waiting for guidance of where/which tests I could do here)

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/14430)
<!-- Reviewable:end -->

julienw committed:
Expose Quirks Mode information in the style shared context r=emilio

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @asajeffrey (or someone else) soon.

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/script/dom/document.rs, components/script/layout_wrapper.rs, components/script_layout_interface/lib.rs, components/script_layout_interface/wrapper_traits.rs, components/script_layout_interface/Cargo.toml
 * @fitzgen: components/script/dom/document.rs, components/script/layout_wrapper.rs, components/script_layout_interface/lib.rs, components/script_layout_interface/wrapper_traits.rs, components/script_layout_interface/Cargo.toml
 * @emilio: components/layout/construct.rs, components/layout/fragment.rs, components/layout/Cargo.toml, components/layout/table_colgroup.rs, components/layout/generated_content.rs, components/layout/block.rs, components/layout/lib.rs, components/layout/inline.rs, components/layout/flow.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify layout and script code, but no tests are modified. Please consider adding a test!

julienw commented on components/layout/fragment.rs:
```diff
@@ -3129,6 +3163,32 @@ bitflags! {
+        const IS_QUIRKS_MODE = 0b0000_0100,
+        /// Which quirks mode we're in: 0 is quirks, 1 is limited quirks
+        const QUIRKS_MODE = 0b0000_1000,
+    }
```
here I decided to integrate it in the bitflags, because I thought fragments are everywhere and I didn't want to use more memory when not needed.

julienw commented on components/layout/generated_content.rs:
```diff
@@ -442,7 +442,8 @@ fn render_text(layout_context: &LayoutContext,
                                                              RestyleDamage::rebuild_and_reflow(),
-                                                             info));
+                                                             info,
+                                                             None));
```
quirksmode information is actually useless here

julienw commented on components/layout/fragment.rs:
```diff
@@ -921,7 +922,34 @@ impl Fragment {
+    pub fn from_node_and_style<N: ThreadSafeLayoutNode>(node: &N,
+                                                        style: Arc<ServoComputedValues>,
+                                                        specific: SpecificFragmentInfo)
+                                                        -> Fragment {
```
I created this to simplify some of the existing code, but it's not _really_ necessary for this patch. Tell me if you'd prefer that I revert the changes around this constructor.

If you decide that we should keep it, maybe `new` could call it too and cut some code.

julienw commented on components/layout/fragment.rs:
```diff
@@ -921,7 +922,34 @@ impl Fragment {
+        let writing_mode = style.writing_mode;
+
+        let mut restyle_damage = node.restyle_damage();
+        restyle_damage.remove(RECONSTRUCT_FLOW);
```
If I read prperly, this single line is usually what is different where I changed the calls from `from_opaque_node_...` to this constructor. I don't know what are the consequences.

emilio commented:
Isn't the Quirks mode a document-level thing? If so, shouldn't it go into `LayoutContext` instead of consuming memory all over the place?

julienw commented:
Yes it is at the document-level; I'll check why I didn't go with this solution.

Note we don't consume more memory if we use the existing bitflags :)

emilio commented:
Yes, but we'd use two flags that would be more helpful for other stuff :)

julienw commented:
Yep, I'll look closer :)
Do you have some ideas about where tests could be added ?

julienw commented:
So I looked closer and now I remember why I did it this way.

Ultimately we need the information in the flow objects (like InlineFlow or BlockFlow) for most of layout-related issues (I especially looked #11704 so far but I expect other issues to have the same requirements).

At first I only added the information in BaseFlow and modified the constructors for InlineFlow and BlockFlow (we don't even need to change LayoutContext for this as we can directly access the data in the document). InlineFlow was quite easy, but I saw that BlockFlow's constructors were used a lot with a fragment as argument. That's when I thought that having this information in a Fragment could be useful as well after all.

Now I understand how LayoutContext works: we create it once for every thread (with some shared data) and we pass it over as parameter to every call. Do you think it would be a better architecture ?

emilio commented:
On Thu, Dec 01, 2016 at 07:10:18AM -0800, Julien Wajsberg wrote:
> Now I understand how LayoutContext works: we create it once for every
> thread (with some shared data) and we pass it over as parameter to
> every call. Do you think it would be a better architecture ?

Yes, I think so, makes sense to put in SharedLayoutContext anything that
is thread-independent and valid for the whole layout pass.

julienw commented:
@emilio I just pushed 2 new commits to revert part of my previous changes and add the information in the shared layout context. Can you tell me what you think ?

emilio commented on components/script/layout_wrapper.rs:
```diff
@@ -908,6 +913,13 @@ impl<'ln> ThreadSafeLayoutNode for ServoThreadSafeLayoutNode<'ln> {
     }
+
+    #[inline]
+    fn quirks_mode(&self) -> QuirksMode {
```
nit: This is not needed anymore, right?

julienw replied:
yeah, I think you're right. If we ever need it later, we can just add it at that time.

emilio commented on components/layout_thread/lib.rs:
```diff
@@ -515,6 +521,7 @@ impl LayoutThread {
                 error_reporter: self.error_reporter.clone(),
                 local_context_creation_data: Mutex::new(local_style_context_creation_data),
                 timer: self.timer.clone(),
+                quirks_mode: self.quirks_mode.unwrap_or(QuirksMode::NoQuirks),
```
I think you should be able to `unwrap` here?

julienw replied:
Looking only at this file, I wasn't sure `quirks_mode` would be always already populated when eg `repaint` or `tick_animations` or others would be called. I think it's not a big deal to have a default value here, what do you think ?

emilio replied:
I think it'd be wrong to lay out a quirks mode document in no-quirks mode, then laying it out in quirks mode, overall taking into account incremental reflows.

julienw replied:
But just unwrapping would panic the thread in the same case; is it wishable  ?

emilio replied:
I think we do. I don't think we can layout anything without a document anyway. Indeed the sanest stuff would be making `build_shared_layout_context` taking the document directly, and don't store the quirks mode in the script thread at all.

emilio commented on components/layout_thread/lib.rs:
```diff
@@ -228,6 +230,9 @@ pub struct LayoutThread {
     // rather limit the dependency on that module here.
     layout_threads: usize,
+
+    /// Which quirks mode are we rendering the document in ?
```
nit: No space before `?`.

julienw replied:
my french typography ;)

emilio reviewed (commented):
Mostly nits, the rest looks sane to me :)

Also, please squash together the commits.

julienw commented:
I fixed your comments except the last one, I squashed the commits.

I'm not sure of what gives the test issue on Travis. I reproduce also locally, but not on master, so it should be something in this patch. But I don't know how to get more information.

SimonSapin commented on components/style/Cargo.toml:
```diff
@@ -30,6 +30,7 @@ euclid = "0.10.1"
 fnv = "1.0"
 heapsize = "0.3.0"
 heapsize_derive = {version = "0.1", optional = true}
+html5ever = {version = "0.10.2", features = ["heap_size", "unstable"]}
```
This is a fairly large dependency to add to Stylo, just to use a small enum. And it is also what causes `build-geckolib` to fail because that uses an older (stable channel) Rust compiler that doesn’t know about new-style procedural macros.

Instead I recommend duplicating the definition of the `QuirksMode` enum into `components/style/context.rs`. For consistency, the rest of Servo should be changed to use that enum instead of `html5ever::tree_builder::QuirksMode`. Namely, these files:

```
components/script/dom/node.rs
components/script/dom/document.rs
components/script/dom/bindings/trace.rs
components/script/dom/servoparser/html.rs
```

The latter has a `set_quirks_mode` method that takes a `html5ever::tree_builder::QuirksMode` and will need to do the conversion to `style::context::QuirksMode`.

Ms2ger commented on Cargo.lock:
```diff
@@ -223,6 +223,13 @@ version = "0.1.17"
 source = "git+https://github.com/browserhtml/browserhtml?branch=crate#aafbb0996b02d1fadd4713c96e3d22b542f175a1"
 
 [[package]]
+name = "build-apk"
```
Revert

julienw replied:
sure, I don't know where this comes from.

SimonSapin replied:
This was a bug unrelated to this PR, it’s fixed on master now. This should disappear if you rebase.

emilio commented:
Does this info need to be in the `SharedStyleContext`? Shouldn't it be better for it to be in the `SharedLayoutContext` if only layout is going to use it?

julienw commented:
> Does this info need to be in the SharedStyleContext? Shouldn't it be better for it to be in the SharedLayoutContext if only layout is going to use it?

Yeah that was also a question I asked to @SimonSapin :) He said that some of the things in the quirks mode are style-related, for example https://quirks.spec.whatwg.org/#the-tables-inherit-color-from-body-quirk

julienw commented on ports/geckolib/glue.rs:
```diff
@@ -115,6 +115,8 @@ fn create_shared_context(mut per_doc_data: &mut AtomicRefMut<PerDocumentStyleDat
         error_reporter: Box::new(StdoutErrorReporter),
         local_context_creation_data: Mutex::new(local_context_data),
         timer: Timer::new(),
+        // FIXME Find the real QuirksMode information for this document
```
do you want a bugzilla bug number for this ?

julienw commented:
Green tests !

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13489) made this pull request unmergeable. Please resolve the merge conflicts.

julienw commented:
> ☔️ The latest upstream changes (presumably #13489) made this pull request unmergeable. Please resolve the merge conflicts. 

done

emilio commented:
Looks good, thanks!

@bors-servo r+

bors-servo commented:
:pushpin: Commit d1f20de has been approved by `emilio`

<!-- @bors-servo r=emilio d1f20de8a504385f9ce5b876a0cdd059bf7b8efe -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #14436) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:lock: Merge conflict

bors-servo commented:
:umbrella: The latest upstream changes (presumably #14610) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
@bors-servo: delegate+
If you rebase again, you can merge it immediately with r=emilio.

bors-servo commented:
:v: @julienw can now approve this pull request

julienw commented:
Just realized how I should add r=emilio :) but afk right now so this will wait a few more hours.

KiChjang commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit d024787 has been approved by `emilio`

<!-- @bors-servo r=emilio d02478718850f5348e203472828ffcf445e5a0ee -->

bors-servo commented:
:hourglass: Testing commit d024787 with merge 164426a...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/4521), [arm32](http://build.servo.org/builders/arm32/builds/4695), [arm64](http://build.servo.org/builders/arm64/builds/4718), [linux-dev](http://build.servo.org/builders/linux-dev/builds/4805), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/1485), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/1488), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4781), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/4819), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/1488), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1523), [windows-dev](http://build.servo.org/builders/windows-dev/builds/2770)

julienw commented:
thanks @kichjang !

emilio commented:
Thank you for the patch @julienw! :)
